File: server/support.go
package main

import "math"

func supportRole(i int) int { return [9]int{0, 1, 1, 2, 2, 2, 4, 4, 3}[i%9] }
func supportTerrain(p Player) (int, int) {
	return int(math.Round(p.Z/(22.4/576) + 320)), int(math.Round(576 - p.X/(22.4/576)))
}
func supportPredicted(p Player) (int, int) {
	x, y := supportTerrain(p)
	return x + int(math.Round(p.moveZ/(22.4/576)/25)), y - int(math.Round(p.moveX/(22.4/576)/25))
}

// base_player_ai / find_closest_available_enemy (0xf786..0xf8ca).
func (s *State) aggressionTarget(i int, distances *[18]int, random int) (float64, float64, bool) {
	p := &s.Players[i]
	if !s.worldInViewport(p.X, p.Z, 0) || random >= p.Stats[0]/2 {
		return 0, 0, false
	}
	side := p.Team
	if s.Period == 2 {
		side ^= 1
	}
	zone := supportZones[side][i%9]
	closest := p.Stats[7] * 2
	var tx, tz float64
	found := false
	for j, q := range s.Players {
		if q.Team == p.Team || q.Stun > 0 || q.Health <= 0 || !s.worldInViewport(q.X, q.Z, 0) || distances[j] >= closest {
			continue
		}
		x, z := predictedTarget(q.X, q.Z, q.moveX, q.moveZ, p.Stats[7])
		a, b := supportTerrain(Player{X: x, Z: z})
		if a < zone[0] || a > zone[1] || b < zone[2] || b > zone[3] {
			continue
		}
		closest, tx, tz, found = distances[j], x, z, true
	}
	return tx, tz, found
}

// Positional branch of base_player_ai, including forward support lookup tables.
func (s *State) supportTarget(i int) (float64, float64) { return s.supportPosition(i, false) }
func (s *State) supportPosition(i int, self bool) (float64, float64) {
	team := i / 9
	side := team
	if s.Period == 2 {
		side ^= 1
	}
	zone := supportZones[side][i%9]
	xmin, xmax, ymin, ymax := zone[0], zone[1], zone[2], zone[3]
	cx, cy := ((xmin+xmax)/2)&^1, ((ymin+ymax)/2)&^1
	r := supportRole(i)
	selected := s.Controlled[team]
	if self {
		selected = i
	}
	if r == 1 && supportRole(selected) >= 2 {
		selected = team * 9
	}
	q := s.Players[selected]
	qr := min(3, supportRole(selected))
	x, y := supportPredicted(q)
	if self {
		x, y = supportTerrain(q)
	}
	qx, _ := supportTerrain(q)
	bound := func(v, a, b int) int { return max(a, min(b, v)) }
	abs := func(v int) int {
		if v < 0 {
			return -v
		}
		return v
	}
	adjust := func() {
		x = bound((x+cx)/2, xmin, xmax)
		delta := abs(x - qx)
		diagonal := false
		if qr <= r {
			if side == 0 {
				diagonal = y >= cy
				if diagonal {
					delta = -delta
				}
			} else {
				diagonal = y <= cy
			}
		} else {
			if side == 0 {
				diagonal = y <= cy
			} else {
				diagonal = y >= cy
				if diagonal {
					delta = -delta
				}
			}
		}
		if diagonal {
			y += delta
			constrained := bound(y, ymin, ymax)
			excess := abs(y - constrained)
			y = constrained
			if x <= cx {
				x += excess
			} else {
				x -= excess
			}
			x = bound(x, xmin, xmax)
		} else {
			_, y = supportPredicted(q)
			y = bound(y, ymin, ymax)
		}
	}
	if supportRole(selected) >= 3 && r >= 3 {
		mx, my := x, y
		if mx >= 320 {
			mx = 639 - mx
		}
		if side != 0 {
			my = 1151 - my
		}
		if my >= 32 && my < 208 && mx >= 32 && mx < 320 {
			index := (my-32)/32*9 + (mx-32)/32
			table := 0
			if r == 4 {
				table = 1
			}
			target := forwardTargets[table][index]
			x, y = target[0], target[1]
			if qx >= 320 {
				x = 639 - x
			}
			if side != 0 {
				y = 1151 - y
			}
			x = bound(x, xmin, xmax)
		} else {
			x, y = supportPredicted(q)
			adjust()
		}
	} else if r != 1 && x > xmin && x < xmax {
		anchor := cy
		if qr > r {
			if side == 0 && y >= cy {
				anchor = ymax
			}
			if side != 0 && y <= cy {
				anchor = ymin
			}
		} else {
			if side == 0 && y <= cy {
				anchor = ymin
			}
			if side != 0 && y >= cy {
				anchor = ymax
			}
		}
		y = bound((y+anchor)/2, ymin, ymax)
	} else {
		adjust()
	}
	return float64(576-y) * (22.4 / 576), float64(x-320) * (22.4 / 576)
}

File: src/support.js
import { zones, forwardTargets } from "./support-data.js";
import { predictedTarget } from "./steering.js";
import { worldInViewport } from "./visibility.js";

const unit = 22.4 / 576;
const role = (i) => [0, 1, 1, 2, 2, 2, 4, 4, 3][i % 9];
const terrain = (p) => [
  Math.round(p.z / unit + 320),
  Math.round(576 - p.x / unit),
];
const predicted = (p) => {
  const [x, y] = terrain(p);
  return [
    x + Math.round((p.moveZ || 0) / unit / 25),
    y - Math.round((p.moveX || 0) / unit / 25),
  ];
};
const clamp = (v, a, b) => Math.max(a, Math.min(b, v));

// base_player_ai / find_closest_available_enemy (0xf786..0xf8ca).
// Reuse the decision's random byte and cached opponent distances.
export function aggressionTarget(s, i, distances, random) {
  const p = s.players[i];
  if (!worldInViewport(s, p) || random >= (p.stats[0] >> 1)) return null;
  const side = p.team ^ (s.period === 2 ? 1 : 0);
  const [xmin, xmax, ymin, ymax] = zones[side][i % 9];
  let closest = p.stats[7] * 2, target = null;
  for (let j = 0; j < s.players.length; j++) {
    const q = s.players[j];
    if (q.team === p.team || q.stun > 0 || q.health <= 0 ||
        !worldInViewport(s, q) || distances[j] >= closest) continue;
    const next = predictedTarget(q.x, q.z, q.moveX || 0, q.moveZ || 0, p.stats[7]);
    const [x, y] = terrain({ x: next[0], z: next[1] });
    if (x < xmin || x > xmax || y < ymin || y > ymax) continue;
    closest = distances[j];
    target = next;
  }
  return target;
}

// Positional branch of base_player_ai: E218..E438 in the C# reference,
// adjust_support_target / attackers_targeting in the Amiga disassembly.
export function supportTarget(s, i, self = false) {
  const team = Math.floor(i / 9),
    side = team ^ (s.period === 2 ? 1 : 0);
  const [xmin, xmax, ymin, ymax] = zones[side][i % 9];
  const cx = Math.floor((xmin + xmax) / 2) & ~1;
  const cy = Math.floor((ymin + ymax) / 2) & ~1;
  const r = role(i);
  let selected = self ? i : s.controlled[team];
  if (r === 1 && role(selected) >= 2) selected = team * 9;
  const q = s.players[selected],
    qr = Math.min(3, role(selected));
  let [x, y] = self ? terrain(q) : predicted(q);
  const [qx] = terrain(q);
  const adjust = () => {
    x = clamp(Math.floor((x + cx) / 2), xmin, xmax);
    let delta = Math.abs(x - qx),
      diagonal;
    if (qr <= r) {
      diagonal = side === 0 ? y >= cy : y <= cy;
      if (side === 0 && diagonal) delta = -delta;
    } else {
      diagonal = side === 0 ? y <= cy : y >= cy;
      if (side !== 0 && diagonal) delta = -delta;
    }
    if (diagonal) {
      y += delta;
      const constrained = clamp(y, ymin, ymax),
        excess = Math.abs(y - constrained);
      y = constrained;
      x = clamp(x + (x <= cx ? excess : -excess), xmin, xmax);
    } else y = clamp(predicted(q)[1], ymin, ymax);
  };
  if (role(selected) >= 3 && r >= 3) {
    const mirrorX = x >= 320 ? 639 - x : x;
    const mirrorY = side === 0 ? y : 1151 - y;
    if (mirrorY >= 32 && mirrorY < 208 && mirrorX >= 32 && mirrorX < 320) {
      const index =
        Math.floor((mirrorY - 32) / 32) * 9 + Math.floor((mirrorX - 32) / 32);
      [x, y] = forwardTargets[r === 4 ? 1 : 0][index];
      if (qx >= 320) x = 639 - x;
      if (side !== 0) y = 1151 - y;
      x = clamp(x, xmin, xmax);
    } else {
      [x, y] = predicted(q);
      adjust();
    }
  } else if (r !== 1 && x > xmin && x < xmax) {
    let anchor;
    if (qr > r)
      anchor = side === 0 ? (y >= cy ? ymax : cy) : y <= cy ? ymin : cy;
    else anchor = side === 0 ? (y <= cy ? ymin : cy) : y >= cy ? ymax : cy;
    y = clamp(Math.floor((y + anchor) / 2), ymin, ymax);
  } else adjust();
  return [(576 - y) * unit, (x - 320) * unit];
}
